ldms/src/sampler/curated_metrics/curated_metrics.c

```diff
@@ -0,0 +1,1147 @@
+	ldms_record_t rec_def;
+	mi->rec_metric_len = 0;
+	for (i = 0; i < metric_list_sz; i++) {
+		if (mi->collect_list[i].collect) {
+			if (metric_list[i].not_rec) {
+				rc = ldms_schema_metric_add(schema, metric_list[i].name, metric_list[i].vtype);
+				if (rc < 0)
+					goto err2;
+			} else {
+				if (!mi->rec_metric_len) {
+					rec_def = ldms_record_create("netdev");
+					if (!rec_def) {
+						rc = errno;
+						goto err2;
+					}
+				}
+				if (metric_list[i].vtype == LDMS_V_CHAR_ARRAY) {
+					rc = ldms_record_metric_add(rec_def, metric_list[i].name,
+						"iface_name", metric_list[i].vtype, IFNAMSIZ);
+				} else {
+					rc = ldms_record_metric_add(rec_def, metric_list[i].name,
+						NULL, metric_list[i].vtype, 0);
+				}
+				mi->rec_metric_len++;
+			}
+		}
+	}
+	/* Begin loadavg metrics */
+	mi->lf = fopen(procloadavg, "r");
+	if (!mi->lf) {
+		ovis_log(mi->log, OVIS_LERROR, SAMP ": Could not open " PROC_LOADAVG
+				" : continuing\n");
+	}
+	fclose(mi->lf);
+	mi->lf = NULL;
+
+	mi->sf = fopen(procstat, "r");
+	if (!mi->sf) {
+		ovis_log(mi->log, OVIS_LERROR, SAMP ": Could not open " PROC_STAT
+				" : continuing\n");
+	}
+	fclose(mi->sf);
+	mi->sf = NULL;
+	/* procnetdev metrics */
+	mi->nf = fopen(netdev, "r");
+	if (!mi->nf) {
+		ovis_log(mi->log, OVIS_LERROR, SAMP ": Could not open " NET_DEV
+				" : continuing\n");
+		rc = ENOENT;
+		goto err;
+	}
+	mi->rec_heap_sz = ldms_record_heap_size_get(rec_def);
+	mi->heap_sz = MAXIFACE * ldms_record_heap_size_get(rec_def);
+
+	mi->rec_def_idx = ldms_schema_record_add(schema, rec_def);
+	if (mi->rec_def_idx < 0) {
+		rc = -mi->rec_def_idx;
+		goto err3;
+	}
+
+	mi->netdev_list_mid = ldms_schema_metric_list_add(schema, "netdev_list", NULL, mi->heap_sz);
+	if (mi->netdev_list_mid < 0) {
+		rc = -mi->netdev_list_mid;
+		goto err2;
+	}
+	fclose(mi->nf);
+	mi->set = base_set_new_heap(mi->base, mi->heap_sz);
+	if (!mi->set) {
+		rc = errno;
+		goto err2;
+	}
+
+	return 0;
+
+err3:
+	ldms_record_delete(rec_def);
+err2:
+	base_schema_delete(mi->base);
+	mi->base = NULL;
+err:
+	if (mi->mf) {
+		fclose(mi->mf);
+		mi->mf = NULL;
+	}
+	if (mi->lf) {
+		fclose(mi->lf);
+		mi->lf = NULL;
+	}
+	if (mi->sf) {
+		fclose(mi->sf);
+		mi->sf = NULL;
+	}
+	if (mi->nf) {
+		fclose(mi->nf);
+		mi->nf = NULL;
+	}
+	return rc;
+}
+
+static const char *usage(ldmsd_plug_handle_t handle)
+{
+	return  "config name=" SAMP " path=<config_filepath>\n" BASE_CONFIG_USAGE \
+		"	<config_filepath>	A path to a json configuration file containing a dictionary with two available keywords.\n"
+		"				The 'metrics' keyword value is a list of metrics in the curated_metrics sampler to include.\n"
+		"				The 'ifaces' keyword value is a list of interfaces to include.\n";
+}
+
+static int config(ldmsd_plug_handle_t handle, struct attr_value_list *kwl, struct attr_value_list *avl)
+{
+	curated_metrics_t mi = ldmsd_plug_ctxt_get(handle);
+	int rc;
+
+	mi->log = ldmsd_plug_log_get(handle);
```

Comment:
It is little better to get a copy of the log handle in a single time in constructor() rather than here in config().